Auth: add IsServiceAccount to IsRealUser

eleijonmarck opened the pull request:
<!--

Thank you for sending a pull request! Here are some tips:

1. If this is your first time, please read our contribution guide at https://github.com/grafana/grafana/blob/main/CONTRIBUTING.md

2. Ensure you include and run the appropriate tests as part of your Pull Request.

3. In a new feature or configuration option, an update to the documentation is necessary. Everything related to the documentation is under the docs folder in the root of the repository.

4. If the Pull Request is a work in progress, make use of GitHub's "Draft PR" feature and mark it as such.

5. If you can not merge your Pull Request due to a merge conflict, Rebase it. This gets it in sync with the main branch.

6. Name your PR as "<FeatureArea>: Describe your change", e.g. Alerting: Prevent race condition. If it's a fix or feature relevant for the changelog describe the user impact in the title. The PR title is used to auto-generate the changelog for issues marked with the "add to changelog" label.

7. If your PR content should be added to the What's New document for the next major or minor release, add the **add to what's new** label to your PR. Note that you should add this label to the main PR that introduces the feature; do not add this label to smaller PRs for the feature.

-->

**What is this feature?**
This adds IsServiceAccount bool ontop of the SIgnedInUser struct to know whether the retrieved user is a service account or not.

**Why do we need this feature?**
We need to in order for us to differentiate between a retrieved user account and a service account account

**Which issue(s) does this PR fix?**:
https://github.com/grafana/grafana/issues/57532

**Special notes for your reviewer**:
Have checked with the original intent of IsRealUser and it is correct to assume that we want to exclude users whom are apikeys or service accounts.
IsRealUser is being used in these places https://github.com/grafana/grafana/search?q=IsRealUser

eleijonmarck committed:
add: IsServiceAccount to SignedInUser and IsRealUser

eleijonmarck commented on pkg/services/ngalert/schedule/schedule.go:
```diff
@@ -328,6 +328,7 @@ func (sch *schedule) ruleRoutine(grafanaCtx context.Context, key ngmodels.AlertR
 		start := sch.clock.Now()
 
 		schedulerUser := &user.SignedInUser{
+			// FIXME: add is service account and refactor to a service account instead of a user
```
@grafana/grafana-edge-squad wanted to add you into the loop to know how to account for grafana service accounts from the signedinuser and that we have added the `IsServiceAccount` attribute to the SignedInUser.

eleijonmarck commented on pkg/services/serviceaccounts/serviceaccounts.go:
```diff
@@ -7,13 +7,29 @@ import (
 	RetrieveServiceAccountIdByName(ctx context.Context, orgID int64, name string) (int64, error)
 }
 
+/*
```
comments outside of the scope of this PR. this comment reflects on the migration of apikeys to service accounts

eleijonmarck commented on pkg/services/user/model.go:
```diff
@@ -203,6 +203,7 @@ type SignedInUser struct {
 	Name               string
 	Email              string
 	ApiKeyID           int64 `xorm:"api_key_id"`
+	IsServiceAccount   bool  `xorm:"is_service_account"`
```
bulk of the change is that we add this parameter

kalleep commented on pkg/services/serviceaccounts/serviceaccounts.go:
```diff
@@ -7,13 +7,29 @@ import (
+// only used for interal api calls
+RevertApiKey reverts a single service account to an API key.
+*/
 type Store interface {
```
The store interface should not be exposed in the domain package. If we want to interact with service accounts the service should be injected. But I think that refactor should be done in another pr

eleijonmarck replied:
il add a new issue for refactoring the store

kalleep commented on pkg/services/user/model.go:
```diff
@@ -276,8 +277,13 @@ func (u *SignedInUser) HasRole(role roletype.RoleType) bool {
+	// checking if userId the user is a real user
+	// previously we used to check if the UserId was 0 or -1
+	// and not a service account
+	return u.UserID > 0 && u.IsServiceAccount == false
```
With this change we need to also update HasUniqueId to return true for service account, otherwise we will never cache permissions for a service account

kalleep reviewed (commented):
Looks good, left a few comments we need to address

toddtreece commented on pkg/services/contexthandler/contexthandler.go:
```diff
@@ -312,7 +312,8 @@ func (h *ContextHandler) initContextWithAPIKey(reqContext *models.ReqContext) bo
 	if apikey.ServiceAccountId == nil || *apikey.ServiceAccountId < 1 { //There is no service account attached to the apikey
-		//Use the old APIkey method.  This provides backwards compatibility.
+		// Use the old APIkey method.  This provides backwards compatibility.
+		// will probably have to be supported for a long time.
```
> // will probably have to be supported for a long time.

i don't have a lot of context here, but are there downsides to making this migration automatic?
<img width="1397" alt="image" src="https://user-images.githubusercontent.com/360020/199482787-d2617588-32f8-4d89-8d5a-46e1b9858828.png">

ryantxu replied:
Or perhaps automatic if no keys exist

eleijonmarck replied:
this is a good take. I'll add this to our backlog

eleijonmarck replied:
@toddtreece alot of our customers rely on apikeys and we want to make sure that they have been aware of the change to service accounts.

Are you referring to a new instance in general, where they do not have any apikeys and we set service accounts tab only by default? 🤔

eleijonmarck replied:
if it's the latter case, I totally agree that that does make sense ⚡

toddtreece replied:
@eleijonmarck i guess i meant either way. i have no context about what actually happens behind the scenes when an API key is migrated using the `Migrate to service accounts now` button, i just see that a service account has been created and the API key still works as it did before migration:
![image](https://user-images.githubusercontent.com/360020/199575134-f3e8cdfb-88ca-4f32-81b1-a0054616a8f5.png)

and then the api keys have this notice:
![image](https://user-images.githubusercontent.com/360020/199576532-a5b36bff-13dc-4413-b96c-e62e05edeb12.png)

eleijonmarck replied:
added a new issue to the backlog - https://github.com/grafana/grafana/issues/58091

eleijonmarck replied:
@toddtreece this is exactly the intended experience. We were the ones developing this workflow, so any feedback is welcome.
We could add some more detail about what is meant by migrating from apikey to service accounts.

toddtreece replied:
nope, no issues with the workflow for migration, it works great! the `// will probably have to be supported for a long time` comment made me wonder if API keys could be automatically migrated at some point so that the logic for the old keys can be removed. a bit of a tangent since i was just curious about the plan, but that can be discussed outside of this PR :smile:

eleijonmarck committed:
fix: linting error

eleijonmarck committed:
Merge branch 'main' into eleijonmarck/serviceaccounts/team-member-bugfix

eleijonmarck committed:
refactor: add function IsServiceAccountUser()

By adding the function IsServiceAccountUser() we use it to identify for
ServiceAccounts in the HasUniqueID() since caching is built up on having
a uniqueID, see comment: https://github.com/grafana/grafana/pull/58015#discussion_r1011361880

eleijonmarck commented on pkg/services/user/model.go:
```diff
@@ -276,16 +277,26 @@ func (u *SignedInUser) HasRole(role roletype.RoleType) bool {
+}
+
 func (u *SignedInUser) HasUniqueId() bool {
-	return u.IsRealUser() || u.IsApiKeyUser()
+	return u.IsRealUser() || u.IsApiKeyUser() || u.IsServiceAccountUser()
 }
```
@kalleep added the method IsServiceAccountUser() on top of the SignedUser struct to be able to use it the same way for apikeyUser and realUser.
This is to account for caching for the service accounts as well 👍
